core/directory.go

```diff
@@ -316,30 +316,117 @@ func (dir *Directory) WithDirectory(ctx context.Context, subdir string, src *Dir
+	})); err != nil {
+		return nil, err
 	}
 
-	st = st.File(llb.Copy(srcSt, src.Dir, path.Join(dir.Dir, subdir), opts...))
+	dir.Services.Merge(src.Services)
 
-	err = dir.SetState(ctx, st)
+	return dir, nil
+}
+
+func (dir *Directory) WithFile(ctx context.Context, destPath string, src *File, permissions fs.FileMode, owner *Ownership) (*Directory, error) {
+	dir = dir.Clone()
+
+	destSt, err := dir.State()
 	if err != nil {
 		return nil, err
 	}
 
+	srcSt, err := src.State()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := dir.SetState(ctx, mergeStates(mergeStateInput{
+		Dest:         destSt,
+		DestDir:      path.Join(dir.Dir, path.Dir(destPath)),
+		DestFileName: path.Base(destPath),
+		Src:          srcSt,
+		SrcDir:       path.Dir(src.File),
+		SrcFileName:  path.Base(src.File),
+		Permissions:  permissions,
+		Owner:        owner,
+	})); err != nil {
+		return nil, err
+	}
+
 	dir.Services.Merge(src.Services)
 
 	return dir, nil
 }
 
+type mergeStateInput struct {
+	Dest         llb.State
+	DestDir      string
+	DestFileName string
+
+	Src         llb.State
+	SrcDir      string
+	SrcFileName string
+
+	IncludePatterns []string
+	ExcludePatterns []string
+
+	Permissions fs.FileMode
+	Owner       *Ownership
+}
+
+func mergeStates(input mergeStateInput) llb.State {
+	input.DestDir = path.Join("/", input.DestDir)
+	input.SrcDir = path.Join("/", input.SrcDir)
+
+	copyInfo := &llb.CopyInfo{
+		CreateDestPath:      true,
+		CopyDirContentsOnly: true,
+		IncludePatterns:     input.IncludePatterns,
+		ExcludePatterns:     input.ExcludePatterns,
+	}
+	if input.DestFileName == "" && input.SrcFileName != "" {
+		input.DestFileName = input.SrcFileName
+	}
+	if input.Permissions != 0 {
+		copyInfo.Mode = &input.Permissions
+	}
+	if input.Owner != nil {
+		input.Owner.Opt().SetCopyOption(copyInfo)
+	}
+
+	// MergeOp currently only supports merging the "/" of states together without any
+	// modifications or filtering
+	canDoDirectMerge := copyInfo.Mode == nil &&
+		copyInfo.ChownOpt == nil &&
+		len(copyInfo.ExcludePatterns) == 0 &&
+		len(copyInfo.IncludePatterns) == 0 &&
+		input.DestDir == input.SrcDir &&
+		input.DestFileName == input.SrcFileName
+
+	mergeStates := []llb.State{input.Dest}
+	if canDoDirectMerge {
+		// Directly merge the states together, which is lazy, uses hardlinks instead of
+		// copies and caches inputs individually instead of invalidating the whole
+		// chain following any modified input.
+		mergeStates = append(mergeStates, input.Src)
+	} else {
+		// Even if we can't merge directly, we can still get some optimization by
+		// copying to scratch and then merging that. This still results in an on-disk
+		// copy but preserves the other caching benefits of MergeOp. This is the same
+		// behavior as "COPY --link" in Dockerfiles.
+		mergeStates = append(mergeStates, llb.Scratch().File(llb.Copy(
+			input.Src, path.Join(input.SrcDir, input.SrcFileName), path.Join(input.DestDir, input.DestFileName), copyInfo,
+		)))
+	}
```

Comment:
I remember being tempted by Merge here but realizing it's more complicated because of the various configs, so it's really cool to see both situations optimized!